File: sdk/go/internal/artifact/step.go
package artifact

import (
	"bytes"
	"fmt"
	"strings"
	"text/template"

	artifactApi "github.com/ALT-F4-LLC/vorpal/sdk/go/api/v0/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/internal/config"
)

type BashScriptTemplateArgs struct {
	Script string
}

type BwrapScriptTemplateArgs struct {
	Script string
}

const BashScriptTemplate = `#!/bin/bash
set -euo pipefail
{{.Script}}
`

const BwrapScriptTemplate = `#!/bin/bash
set -euo pipefail
{{.Script}}
`

func Bash(
	context *config.ConfigContext,
	artifacts []*string,
	environments []string,
	script *string,
	systems []artifactApi.ArtifactSystem,
) (*artifactApi.ArtifactStep, error) {
	stepEnvironments := make([]string, 0)

	stepEntrypoint := "bash"

	for _, value := range environments {
		if strings.Contains(value, "PATH=") {
			continue
		}

		stepEnvironments = append(stepEnvironments, value)
	}

	stepPathBins := make([]string, 0)

	for _, art := range artifacts {
		stepPathBins = append(stepPathBins, fmt.Sprintf("%s/bin", GetEnvKey(art)))
	}

	stepPathDefault := "/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin"

	stepPath := fmt.Sprintf("%s:%s", strings.Join(stepPathBins, ":"), stepPathDefault)

	for _, value := range environments {
		if strings.Contains(value, "PATH=") {
			stepPath = fmt.Sprintf("%s:%s", strings.ReplaceAll(value, "PATH=", ""), stepPath)
		}
	}

	stepEnvironments = append(stepEnvironments, fmt.Sprintf("PATH=%s", stepPath))

	scriptTemplate, err := template.New("script").Parse(BashScriptTemplate)
	if err != nil {
		return nil, err
	}

	var scriptBuffer bytes.Buffer

	scriptTemplateVars := BashScriptTemplateArgs{
		Script: *script,
	}

	if err := scriptTemplate.Execute(&scriptBuffer, scriptTemplateVars); err != nil {
		return nil, err
	}

	step := NewArtifactStepBuilder()

	step = step.WithArtifacts(artifacts, systems)
	step = step.WithEntrypoint(stepEntrypoint, systems)
	step = step.WithEnvironments(stepEnvironments, systems)
	step = step.WithScript(scriptBuffer.String(), systems)

	return step.Build(context), nil
}

func Bwrap(
	context *config.ConfigContext,
	arguments []string,
	artifacts []*string,
	environments []string,
	rootfs *string,
	script string,
	systems []artifactApi.ArtifactSystem,
) (*artifactApi.ArtifactStep, error) {
	// Setup arguments

	stepArguments := []string{
		"--unshare-all",
		"--share-net",
		"--clearenv",
		"--chdir",
		"$VORPAL_WORKSPACE",
		"--gid",
		"1000",
		"--uid",
		"1000",
		"--dev",
		"/dev",
		"--proc",
		"/proc",
		"--tmpfs",
		"/tmp",
		"--bind",
		"$VORPAL_OUTPUT",
		"$VORPAL_OUTPUT",
		"--bind",
		"$VORPAL_WORKSPACE",
		"$VORPAL_WORKSPACE",
		"--setenv",
		"VORPAL_OUTPUT",
		"$VORPAL_OUTPUT",
		"--setenv",
		"VORPAL_WORKSPACE",
		"$VORPAL_WORKSPACE",
	}

	// Setup artifacts arguments

	stepArtifacts := make([]*string, 0)

	if rootfs != nil {
		rootfsArgs := []string{
			"--ro-bind",
			fmt.Sprintf("%s/bin", GetEnvKey(rootfs)),
			"/bin",
			"--ro-bind",
			fmt.Sprintf("%s/etc", GetEnvKey(rootfs)),
			"/etc",
			"--ro-bind",
			fmt.Sprintf("%s/lib", GetEnvKey(rootfs)),
			"/lib",
			"--ro-bind-try",
			fmt.Sprintf("%s/lib64", GetEnvKey(rootfs)),
			"/lib64",
			"--ro-bind",
			fmt.Sprintf("%s/sbin", GetEnvKey(rootfs)),
			"/sbin",
			"--ro-bind",
			fmt.Sprintf("%s/usr", GetEnvKey(rootfs)),
			"/usr",
		}

		stepArguments = append(stepArguments, rootfsArgs...)
		stepArtifacts = append(stepArtifacts, rootfs)
	}

	for _, artifact := range artifacts {
		stepArtifacts = append(stepArtifacts, artifact)
	}

	for _, art := range stepArtifacts {
		stepArguments = append(stepArguments, "--ro-bind")
		stepArguments = append(stepArguments, GetEnvKey(art))
		stepArguments = append(stepArguments, GetEnvKey(art))
		stepArguments = append(stepArguments, "--setenv")
		stepArguments = append(stepArguments, strings.ReplaceAll(GetEnvKey(art), "$", ""))
		stepArguments = append(stepArguments, GetEnvKey(art))
	}

	// Setup environment arguments

	stepPathBins := make([]string, 0)

	for _, art := range stepArtifacts {
		stepPathBins = append(stepPathBins, fmt.Sprintf("%s/bin", GetEnvKey(art)))
	}

	stepPath := fmt.Sprintf("%s:%s", strings.Join(stepPathBins, ":"), "/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin")

	for _, envvar := range environments {
		if strings.Contains(envvar, "PATH=") {
			stepPath = fmt.Sprintf("%s:%s", strings.ReplaceAll(envvar, "PATH=", ""), stepPath)
		}
	}

	stepArguments = append(stepArguments, "--setenv")
	stepArguments = append(stepArguments, "PATH")
	stepArguments = append(stepArguments, stepPath)

	for _, envvar := range environments {
		key := strings.Split(envvar, "=")[0]
		value := strings.Split(envvar, "=")[1]

		if strings.Contains(key, "PATH") {
			continue
		}

		stepArguments = append(stepArguments, "--setenv")
		stepArguments = append(stepArguments, key)
		stepArguments = append(stepArguments, value)
	}

	// Setup arguments

	for _, argument := range arguments {
		stepArguments = append(stepArguments, argument)
	}

	// Setup script

	scriptTemplate, err := template.New("script").Parse(BwrapScriptTemplate)
	if err != nil {
		return nil, err
	}

	var scriptBuffer bytes.Buffer

	scriptTemplateVars := BwrapScriptTemplateArgs{
		Script: script,
	}

	if err := scriptTemplate.Execute(&scriptBuffer, scriptTemplateVars); err != nil {
		return nil, err
	}

	// Setup step

	step := NewArtifactStepBuilder()

	step = step.WithArguments(stepArguments, systems)
	step = step.WithArtifacts(stepArtifacts, systems)
	step = step.WithEntrypoint("bwrap", systems)
	step = step.WithEnvironments([]string{"PATH=/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin"}, systems)
	step = step.WithScript(scriptBuffer.String(), systems)

	return step.Build(context), nil
}

func Shell(
	context *config.ConfigContext,
	artifacts []*string,
	environments []string,
	script string,
) (*artifactApi.ArtifactStep, error) {
	stepTarget := context.GetTarget()

	if stepTarget == artifactApi.ArtifactSystem_AARCH64_DARWIN || stepTarget == artifactApi.ArtifactSystem_X8664_DARWIN {
		return Bash(
			context,
			artifacts,
			environments,
			&script,
			[]artifactApi.ArtifactSystem{
				artifactApi.ArtifactSystem_AARCH64_DARWIN,
				artifactApi.ArtifactSystem_X8664_DARWIN,
			},
		)
	}

	if stepTarget == artifactApi.ArtifactSystem_AARCH64_LINUX || stepTarget == artifactApi.ArtifactSystem_X8664_LINUX {
		linux_vorpal, err := context.FetchArtifact("79958083229520a9e8cbf94ed5f6da40c3bf98ae666b3e20c01bc272fd92c2bb")
		if err != nil {
			return nil, err
		}

		return Bwrap(
			context,
			[]string{},
			artifacts,
			environments,
			linux_vorpal,
			script,
			[]artifactApi.ArtifactSystem{
				artifactApi.ArtifactSystem_AARCH64_LINUX,
				artifactApi.ArtifactSystem_X8664_LINUX,
			},
		)
	}

	return nil, fmt.Errorf("unsupported shell step system: %s", stepTarget)
}

func Docker(
	context *config.ConfigContext,
	arguments []string,
	artifacts []*string,
	systems []artifactApi.ArtifactSystem,
) *artifactApi.ArtifactStep {
	step := NewArtifactStepBuilder()

	step = step.WithArguments(arguments, systems)
	step = step.WithArtifacts(artifacts, systems)
	step = step.WithEntrypoint("docker", systems)
	step = step.WithEnvironments([]string{"PATH=/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin"}, systems)

	return step.Build(context)
}

File: crates/sdk/src/artifact/step.rs
use crate::{
    artifact::{get_env_key, linux_vorpal, ArtifactStepBuilder},
    context::ConfigContext,
};
use anyhow::{bail, Result};
use indoc::formatdoc;
use vorpal_schema::artifact::v0::{
    ArtifactStep, ArtifactSystem,
    ArtifactSystem::{Aarch64Darwin, Aarch64Linux, X8664Darwin, X8664Linux},
};

// TODO: implement amber step

pub fn bash(
    context: &mut ConfigContext,
    artifacts: Vec<String>,
    environments: Vec<String>,
    script: String,
    systems: Vec<ArtifactSystem>,
) -> ArtifactStep {
    let mut step_environments = vec![];

    for environment in environments.iter() {
        if environment.starts_with("PATH=") {
            continue;
        }

        step_environments.push(environment.to_string());
    }

    let step_path_bins = artifacts
        .iter()
        .map(|a| format!("{}/bin", get_env_key(a)))
        .collect::<Vec<String>>()
        .join(":");

    let step_path_default = "/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin";

    let mut step_path = format!("{step_path_bins}:{step_path_default}");

    if let Some(path) = environments.iter().find(|x| x.starts_with("PATH=")) {
        if let Some(path_value) = path.split('=').nth(1) {
            step_path = format!("{path_value}:{step_path}");
        }
    }

    step_environments.push(format!("PATH={}", step_path));

    let step_script = formatdoc! {"
        #!/bin/bash
        set -euo pipefail

        {script}
    "};

    ArtifactStepBuilder::new()
        .with_artifacts(artifacts, systems.clone())
        .with_entrypoint("bash", systems.clone())
        .with_environments(step_environments, systems.clone())
        .with_script(step_script, systems)
        .build(context)
}

pub async fn bwrap(
    context: &mut ConfigContext,
    arguments: Vec<&str>,
    artifacts: Vec<String>,
    environments: Vec<String>,
    rootfs: Option<String>,
    script: String,
    systems: Vec<ArtifactSystem>,
) -> Result<ArtifactStep> {
    // Setup arguments

    let mut step_arguments = vec![
        "--unshare-all".to_string(),
        "--share-net".to_string(),
        "--clearenv".to_string(),
        "--chdir".to_string(),
        "$VORPAL_WORKSPACE".to_string(),
        "--gid".to_string(),
        "1000".to_string(),
        "--uid".to_string(),
        "1000".to_string(),
        "--dev".to_string(),
        "/dev".to_string(),
        "--proc".to_string(),
        "/proc".to_string(),
        "--tmpfs".to_string(),
        "/tmp".to_string(),
        "--bind".to_string(),
        "$VORPAL_OUTPUT".to_string(),
        "$VORPAL_OUTPUT".to_string(),
        "--bind".to_string(),
        "$VORPAL_WORKSPACE".to_string(),
        "$VORPAL_WORKSPACE".to_string(),
        "--setenv".to_string(),
        "VORPAL_OUTPUT".to_string(),
        "$VORPAL_OUTPUT".to_string(),
        "--setenv".to_string(),
        "VORPAL_WORKSPACE".to_string(),
        "$VORPAL_WORKSPACE".to_string(),
    ];

    // Setup artifacts

    let mut step_artifacts = vec![];

    if let Some(rootfs) = rootfs {
        let rootfs_env = get_env_key(&rootfs);
        let rootfs_bin = format!("{rootfs_env}/bin");
        let rootfs_etc = format!("{rootfs_env}/etc");
        let rootfs_lib = format!("{rootfs_env}/lib");
        let rootfs_lib64 = format!("{rootfs_env}/lib64");
        let rootfs_sbin = format!("{rootfs_env}/sbin");
        let rootfs_usr = format!("{rootfs_env}/usr");

        let rootfs_args = vec![
            "--ro-bind".to_string(),
            rootfs_bin,
            "/bin".to_string(),
            "--ro-bind".to_string(),
            rootfs_etc,
            "/etc".to_string(),
            "--ro-bind".to_string(),
            rootfs_lib,
            "/lib".to_string(),
            "--ro-bind-try".to_string(),
            rootfs_lib64,
            "/lib64".to_string(),
            "--ro-bind".to_string(),
            rootfs_sbin,
            "/sbin".to_string(),
            "--ro-bind".to_string(),
            rootfs_usr,
            "/usr".to_string(),
        ];

        step_arguments.extend(rootfs_args);
        step_artifacts.push(rootfs);
    }

    // Setup artifact arguments

    for artifact in artifacts.into_iter() {
        step_artifacts.push(artifact);
    }

    for artifact in step_artifacts.iter() {
        step_arguments.push("--ro-bind".to_string());
        step_arguments.push(get_env_key(artifact));
        step_arguments.push(get_env_key(artifact));
        step_arguments.push("--setenv".to_string());
        step_arguments.push(get_env_key(artifact).replace("$", ""));
        step_arguments.push(get_env_key(artifact));
    }

    // Setup environment arguments

    let step_path_bins = step_artifacts
        .iter()
        .map(|a| format!("{}/bin", get_env_key(a)))
        .collect::<Vec<String>>()
        .join(":");

    let mut step_path = format!("{step_path_bins}:/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin");

    if let Some(path) = environments.iter().find(|x| x.starts_with("PATH=")) {
        if let Some(path_value) = path.split('=').nth(1) {
            step_path = format!("{}:{}", path_value, step_path);
        }
    }

    step_arguments.push("--setenv".to_string());
    step_arguments.push("PATH".to_string());
    step_arguments.push(step_path);

    for env in environments.iter() {
        let key = env.split("=").next().unwrap();
        let value = env.split("=").last().unwrap();

        if key.starts_with("PATH") {
            continue;
        }

        step_arguments.push("--setenv".to_string());
        step_arguments.push(key.to_string());
        step_arguments.push(value.to_string());
    }

    // Setup arguments

    for argument in arguments.into_iter() {
        step_arguments.push(argument.to_string());
    }

    // Setup script

    let step_script = formatdoc! {"
        #!/bin/bash
        set -euo pipefail

        {script}
    "};

    // Setup step

    let step = ArtifactStepBuilder::new()
        .with_arguments(
            step_arguments.iter().map(|x| x.as_str()).collect(),
            systems.clone(),
        )
        .with_artifacts(step_artifacts, systems.clone())
        .with_entrypoint("bwrap", systems.clone())
        .with_environments(
            vec!["PATH=/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin".to_string()],
            systems.clone(),
        )
        .with_script(step_script, systems)
        .build(context);

    Ok(step)
}

pub async fn shell(
    context: &mut ConfigContext,
    artifacts: Vec<String>,
    environments: Vec<String>,
    script: String,
) -> Result<ArtifactStep> {
    // Setup target

    let step_target = context.get_target();

    // Setup step

    let step = match step_target {
        Aarch64Darwin | X8664Darwin => bash(
            context,
            artifacts,
            environments.clone(),
            script.to_string(),
            vec![Aarch64Darwin, X8664Darwin],
        ),

        Aarch64Linux | X8664Linux => {
            let linux_vorpal = linux_vorpal::build(context).await?;

            bwrap(
                context,
                vec![],
                artifacts,
                environments,
                Some(linux_vorpal),
                script,
                vec![Aarch64Linux, X8664Linux],
            )
            .await?
        }

        _ => bail!(
            "unsupported shell step system: {}",
            step_target.as_str_name()
        ),
    };

    Ok(step)
}

pub fn docker(
    context: &mut ConfigContext,
    arguments: Vec<&str>,
    artifacts: Vec<String>,
    systems: Vec<ArtifactSystem>,
) -> ArtifactStep {
    ArtifactStepBuilder::new()
        .with_arguments(arguments, systems.clone())
        .with_artifacts(artifacts, systems.clone())
        .with_entrypoint("docker", systems.clone())
        .with_environments(
            vec!["PATH=/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin".to_string()],
            systems.clone(),
        )
        .build(context)
}
